query/src/main/java/org/infinispan/query/backend/QueryInterceptor.java

```diff
@@ -145,77 +145,79 @@ private Object handleDataWriteCommand(InvocationContext ctx, DataWriteCommand co
          return invokeNext(ctx, command);
       }
-      CacheEntry entry = ctx.lookupEntry(command.getKey());
-      if (ctx.isInTxScope()) {
-         // replay of modifications on remote node by EntryWrappingVisitor
-         if (entry != null && !entry.isChanged() && (entry.getValue() != null || !unreliablePreviousValue(command))) {
-            Map<Object, Object> oldValues = registerOldValues((TxInvocationContext) ctx);
-            oldValues.putIfAbsent(command.getKey(), entry.getValue());
+      return invokeNextThenApply(ctx, command, (rCtx, cmd, rv) -> {
+         if (!cmd.isSuccessful()) {
+            return rv;
          }
+         boolean unreliablePrevious = unreliablePreviousValue(command);
+         if (ctx.isInTxScope()) {
+            Map<Object, Object> oldValues = getOldValuesMap((TxInvocationContext<?>) ctx);
+            registerOldValue(ctx, command.getKey(), unreliablePrevious, oldValues);
+         } else {
+            // TODO: need to reduce the scope of the blocking thread to less if possible later as part of
+            // https://issues.redhat.com/browse/ISPN-11731
+            return asyncValue(blockingManager.runBlocking(() -> indexIfNeeded(rCtx, cmd, unreliablePrevious, cmd.getKey()), cmd)
+                  .thenApply(ignore -> rv));
+         }
+         return rv;
+      });
+   }
+
+   private Object handleManyWriteCommand(InvocationContext ctx, WriteCommand command) {
+      if (command.hasAnyFlag(FlagBitSets.SKIP_INDEXING)) {
          return invokeNext(ctx, command);
-      } else {
-         Object prev = entry != null ? entry.getValue() : UNKNOWN;
-         if (prev == null && unreliablePreviousValue(command)) {
-            prev = UNKNOWN;
+      }
+      return invokeNextThenAccept(ctx, command, (rCtx, cmd, rv) -> {
+         if (!cmd.isSuccessful()) {
+            return;
          }
-         Object oldValue = prev;
-         return invokeNextThenApply(ctx, command, (rCtx, cmd, rv) -> {
-            if (!cmd.isSuccessful()) {
-               return rv;
+         boolean unreliablePrevious = unreliablePreviousValue(command);
+         if (ctx.isInTxScope()) {
+            Map<Object, Object> oldValues = getOldValuesMap((TxInvocationContext<?>) ctx);
+            for (Object key : command.getAffectedKeys()) {
+               registerOldValue(ctx, key, unreliablePrevious, oldValues);
             }
-            CacheEntry entry2 = entry != null ? entry : rCtx.lookupEntry(cmd.getKey());
-            if (entry2 != null && entry2.isChanged()) {
-               // TODO: need to reduce the scope of the blocking thread to less if possible later as part of
-               // https://issues.redhat.com/browse/ISPN-11731
-               return asyncValue(processChange(rCtx, cmd, cmd.getKey(), oldValue, entry2.getValue())
-                     .thenApply(ignore -> rv));
+         } else {
+            for (Object key : cmd.getAffectedKeys()) {
+               indexIfNeeded(rCtx, cmd, unreliablePrevious, key);
```

Comment:
actually, you aren't waiting for the index to complete. maybe it is the source of your issues? I just notice you never wait on `processChange()` `CompletableFuture`

Reply:
yes, it's not waiting here and in the other place it is called. I fix it yesterday but still getting errors, I am debugging now, thanks for the help @pruivo